File: app/src/main/java/com/guanhaolin/pearch/adapter/ImageAdapter.java
package com.guanhaolin.pearch.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.guanhaolin.pearch.api.model.ImageResponse;
import com.guanhaolin.pearch.databinding.ImageViewHolderBinding;
import java.util.ArrayList;
import java.util.List;

public class ImageAdapter
    extends androidx.recyclerview.widget.RecyclerView.Adapter<
        androidx.recyclerview.widget.RecyclerView.ViewHolder> {

  public interface OnImageCellClickListener {
    void onImageCellClicked(ImageResponse image);
  }

  private final List<ImageResponse> images = new ArrayList<>();

  private OnImageCellClickListener onImageCellClickListener;

  private final View.OnClickListener onClickListener =
      new View.OnClickListener() {
        @Override
        public void onClick(View view) {
          if (onImageCellClickListener != null) {
            ImageResponse image = (ImageResponse) view.getTag();
            if (image != null) {
              onImageCellClickListener.onImageCellClicked(image);
            }
          }
        }
      };

  public void setOnImageCellClickListener(OnImageCellClickListener onImageCellClickListener) {
    this.onImageCellClickListener = onImageCellClickListener;
  }

  public void update(List<ImageResponse> newPhotos) {
    images.clear();
    images.addAll(newPhotos);
    notifyDataSetChanged();
  }

  public void clear() {
    images.clear();
    notifyDataSetChanged();
  }

  @NonNull
  @Override
  public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    ImageViewHolderBinding viewbinding =
        ImageViewHolderBinding.inflate(LayoutInflater.from(parent.getContext()));
    viewbinding.root.setOnClickListener(onClickListener);
    return new ImageViewHolder(viewbinding);
  }

  @Override
  public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
    ImageResponse image = images.get(position);
    ImageViewHolder viewHolder = (ImageViewHolder) holder;
    viewHolder.bind(image);
  }

  @Override
  public int getItemCount() {
    return images.size();
  }
}

File: app/src/main/java/com/guanhaolin/pearch/adapter/ImageAdapter.kt
package com.guanhaolin.pearch.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.guanhaolin.pearch.api.model.ImageResponse
import com.guanhaolin.pearch.databinding.ImageViewHolderBinding

class ImageAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    interface OnImageCellClickListener {
        fun onImageCellClicked(image: ImageResponse)
    }

    private val images: MutableList<ImageResponse> = ArrayList()

    private var onImageCellClickListener: OnImageCellClickListener? = null

    private val onClickListener =
        View.OnClickListener { view ->
            val image = view.tag as ImageResponse
            onImageCellClickListener?.onImageCellClicked(image)
        }

    fun setOnImageCellClickListener(onImageCellClickListener: OnImageCellClickListener?) {
        this.onImageCellClickListener = onImageCellClickListener
    }

    fun update(newPhotos: List<ImageResponse>) {
        images.clear()
        images.addAll(newPhotos)
        notifyDataSetChanged()
    }

    fun clear() {
        images.clear()
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val viewbinding =
            ImageViewHolderBinding.inflate(LayoutInflater.from(parent.context))
        viewbinding.root.setOnClickListener(onClickListener)
        return ImageViewHolder(viewbinding)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val image = images[position]
        val viewHolder = holder as ImageViewHolder
        viewHolder.bind(image)
    }

    override fun getItemCount(): Int {
        return images.size
    }
}
